src/schema/create-table-builder.ts

```diff
@@ -263,8 +283,22 @@ export class CreateTableBuilder<TB extends string, C extends string = never>
     columns: C[],
     targetTable: string,
     targetColumns: string[],
-    build: ForeignKeyConstraintBuilderCallback = noop
-  ): CreateTableBuilder<TB, C> {
+    build?: ForeignKeyConstraintBuilderCallback
+  ): CreateTableBuilder<TB, C>
+  addForeignKeyConstraint(
+    columns: C[],
+    targetTable: string,
+    targetColumns: string[],
+    build?: ForeignKeyConstraintBuilderCallback
+  ): CreateTableBuilder<TB, C>
+  addForeignKeyConstraint(...args: any[]): CreateTableBuilder<TB, C> {
+    const constraintName = typeof args[0] === 'string' ? args[0] : undefined
+    const columns: C[] = typeof args[0] === 'string' ? args[1] : args[0]
+    const targetTable: string = typeof args[0] === 'string' ? args[2] : args[1]
+    const targetColumns: string[] =
+      typeof args[0] === 'string' ? args[3] : args[2]
+    const build: ForeignKeyConstraintBuilderCallback =
+      (typeof args[0] === 'string' ? args[4] : args[3]) ?? noop
```

Comment:
Even more painful.

Reply:
How about changing the signature completely (the PR is anyway marked as label:breaking)? For example,

```typescript
addForeignConstraint(
  columns: C[],
  targetTable: string,
  targetColumns: string[],
  build: ForeignKeyConstraintBuilderCallback = noop,
  constraintName?: string,
)
```

or even

```typescript
addForeignConstraint(opts: {
  columns: C[]
  targetTable: string
  targetColumns: string[]
  build?: ForeignKeyConstraintBuilderCallback
  constraintName?: string
})
```
?

Reply:
We're trying to be as close as possible to SQL, JSON inputs are distant from that.
I believe we should add `.references(table: string, columns: string[])` to `ForeignKeyConstraintBuilderInterface`,
and make the FK constraint builder callback a required argument.

So the problematic functions look like:

```ts
db.schema.createTable('table1').addForeignKeyConstraint(
  'constraintName',
  ['col1'],
  (qb) => qb.references('table2', ['col1'])
)
```

```ts
db.schema.createTable('table1').addForeignKeyConstraint(
  ['col1'],
  (qb) => qb.references('table2', ['col1'])
)
```

@koskimas WDYT?

Reply:
The `references` clause is required in SQL (unlike `on delete`/`on update`), so when using it in a builder, we'll need some additional types for builder where **not** calling `.references` on it, the compilation shall fail. I'd lean towards optional builder and...
```ts
db.schema.alterTable('table1').addForeignKeyConstraint(
  ['table1_col'],
  'table2',
  ['table2_col'],
  qb => qb.constraintName('constraintName'), /* .onDelete(...) .onUpdate(...) etc */
)
```

Reply:
> The references clause is required in SQL, unlike on delete/on update (at least in postgres), so when using it in a builder, we'll need some additional types on its methods where not calling .references shall fail the compilation.

We've done it before, most common is requiring aliased builders.
The callback's argument is a builder with nothing but a `.references` method.
The callback's expected return type is a different builder that is returned from `.references` invocation.

The API would be much more readable this way.

> I'd lean towards optional builder and...

Yeah losing the overload is a plus.
But the order of SQL is not aligned with the order of reading/writing this invocation, which is not intuitive.